interp/src/interpreter/steppers/component_interpreter.rs

```diff
@@ -1,88 +1,367 @@
+    for StructuralOrControl<'a, 'outer>
+{
+    fn from(input: ControlInterpreter<'a, 'outer>) -> Self {
         Self::Control(input)
     }
 }
 
-pub struct ComponentInterpreter<'a> {
-    interp: StructuralOrControl<'a>,
+pub struct ComponentInterpreter<'a, 'outer> {
+    interp: StructuralOrControl<'a, 'outer>,
+    input_ports: Vec<RRC<Port>>,
+    output_ports: Vec<RRC<Port>>,
+    comp_ref: &'a Component,
+    control_ref: &'a ir::Control,
+    done_port: RRC<Port>,
+    go_port: RRC<Port>,
+    input_hash_set: Rc<HashSet<*const ir::Port>>,
 }
 
-impl<'a> ComponentInterpreter<'a> {
+impl<'a, 'outer> ComponentInterpreter<'a, 'outer> {
     pub fn from_component(
         comp: &'a Component,
         control: &'a ir::Control,
-        env: InterpreterState,
+        env: InterpreterState<'outer>,
     ) -> Self {
+        let (mut inputs, mut outputs) = (Vec::new(), Vec::new());
+
+        for port in comp.signature.borrow().ports.iter() {
+            let pt_ref = port.borrow();
+            match &pt_ref.direction {
+                ir::Direction::Input => outputs.push(port.clone()),
+                ir::Direction::Output => inputs.push(port.clone()),
+                ir::Direction::Inout => {
+                    panic!()
+                    // empty for now also probably shouldn't happen
+                }
+            }
+        }
+
+        let go_port = inputs
+            .iter()
+            .find(|x| x.borrow().attributes.has("go"))
+            .unwrap()
+            .clone();
+
+        let done_port = outputs
+            .iter()
+            .find(|x| x.borrow().attributes.has("done"))
+            .unwrap()
+            .clone();
+
+        let mut input_hash_set =
+            inputs.iter().map(|x| x.as_raw()).collect::<HashSet<_>>();
+        input_hash_set.insert(done_port.as_raw());
+
+        let input_hash_set = Rc::new(input_hash_set);
         let interp;
 
         if control_is_empty(control) {
             interp = StructuralInterpreter::from_component(comp, env).into();
         } else {
-            interp = ControlInterpreter::new(
-                control,
-                env,
-                &comp.continuous_assignments,
-            )
-            .into()
+            interp = StructuralOrControl::Env(env);
         };
 
-        Self { interp }
+        Self {
+            interp,
+            input_ports: inputs,
+            output_ports: outputs,
+            comp_ref: comp,
+            control_ref: control,
+            go_port,
+            done_port,
+            input_hash_set,
+        }
+    }
+
+    fn look_up_outputs(&self) -> Vec<(ir::Id, crate::values::Value)> {
+        let env = self.get_env();
+        self.output_ports
+            .iter()
+            .map(|x| {
+                let port_ref = x.borrow();
+                (port_ref.name.clone(), env.lookup(x.as_raw()).clone())
+            })
+            .collect()
+    }
+
+    fn go_is_high(&self) -> bool {
```

Comment:
Worth adding the `#[inline]` directive on the next six functions.

Reply:
👍🏻